# SIG720: Machine Learning 

# TASK 6:

## B PRIYANKAA

### s224207694
### priyayj2016@gmail.com

## TARGET GRADE: PASS(P)

## ABSTRACT:

The MNIST digits classification dataset is a well-known dataset in the field of machine learning, particularly in image processing and pattern recognition. It is often used as a benchmark for training and testing algorithms for image classification.

## ABOUT THE DATASET:

The dataset contains 70,000 grayscale images of handwritten digits from 0 to 9. Each image is a 28x28 pixel grid, making a total of 784 pixels. Each image is associated with a label corresponding to the digit it represents, ranging from 0 to 9.

### The dataset is divided into two parts:

- Training Set: 60,000 images used to train machine learning models.
- Test Set: 10,000 images used to evaluate the performance of the trained model.

## TASKS:

1. To load the MNIST Digit dataset and show the size of the training and test set.
2. To develop a one hidden layer multi-layer perceptron model on the above training data and report the accuracy of the model.
3. To set the number of hidden layers of the MLP model as [2,4,6,8,10] and to set the hidden layer size as 100, show the accuracies on the test set.
4. To set the hidden layer size as [50, 100, 150, 200] and to show the accuracies on the test set for the hidden layer.
5. To explain the key findings.

### All the necessary libraries are imported.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import warnings

from scipy.stats import zscore

from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

warnings.filterwarnings('ignore')
%matplotlib inline

# 1.Load the MNIST Digit dataset, show the size of the training and test set.

### Step 1:

1. To load the **mnist** dataset into the variable, "mnist_df". 
2. To set the target variable of mnist_data into y and data columns of mnist_data is set into X.
3. To view the size of the training and test set.

In [2]:
# Load the MNIST dataset

mnist = fetch_openml('mnist_784', version=1)
X, y = mnist["data"], mnist["target"]

In [3]:
X.head()

pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  pixel9  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   pixel10  ...  pixel775  pixel776  pixel777  pixel778  pixel779  pixel780  \
0        0  ...         0         0         0         0         0         0   
1        0  ...         0         0         0         0         0         0   
2        0  ...         0         0         0         0         0         0   
3        0  ...         0         0         0         0         0         0   
4        0  ...         0         0         0         0         0         0   

   pixel781  pixel782  pixel783  pixel784  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 784 columns]

### There are totally 70000 entries in the mnist_df dataset.

In [4]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70000 entries, 0 to 69999
Columns: 784 entries, pixel1 to pixel784
dtypes: int64(784)
memory usage: 418.7 MB


In [5]:
# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Show the size of the training and test set
print(f"Training set size: {X_train.shape[0]}")
print(f"Test set size: {X_test.shape[0]}")
print(f"Training set size: {y_train.shape[0]}")
print(f"Test set size: {y_test.shape[0]}")

Training set size: 56000
Test set size: 14000
Training set size: 56000
Test set size: 14000


### The training dataset for X_train, y_train is 56000 entries. The testing dataset for X_test, y_test is 14000 entries. 

# 2. Develop a one hidden layer multi-layer perceptron model on the above training data, report the accuracy of the model.


## MLP CLASSIFIER:

An artificial neural network used in machine learning for supervised learning tasks such as classification and regression is called as Multi-Layer Perceptron (MLP). It is one of the foundational models in neural networks and deep learning. The input layer consists of nodes (or neurons) that represent the features of the input data. Each node corresponds to one feature of the input data. MLPs have one or more hidden layers between the input and output layers.
Each neuron in a hidden layer applies a weighted sum of its inputs, passes it through an activation function, and produces an output that becomes the input to the next layer. The output layer consists of neurons that represent the possible classes or the output values.

### Step 2:

1. Here we need to Develop a one hidden layer multi-layer perceptron model (MLP) on the training dataset.
2. To report the accuracy of the model.

### MLP Classifier has several parameters.

MLPClassifier(hidden_layer_sizes=(100,), activation='relu', *, solver='adam', alpha=0.0001, batch_size='auto', learning_rate='constant', learning_rate_init=0.001, power_t=0.5, max_iter=200, shuffle=True, random_state=None, tol=0.0001, verbose=False, warm_start=False, momentum=0.9, nesterovs_momentum=True, early_stopping=False, validation_fraction=0.1, beta_1=0.9, beta_2=0.999, epsilon=1e-08, n_iter_no_change=10, max_fun=15000)

#### Here the max_iter is changed to 300 and random_state is set to 42. Other parameters are set to their default values.

### Here the number of layers, n_layers - 2 and hidden_layer_sizes is set to 100.

In [6]:
mlp = MLPClassifier(hidden_layer_sizes=(100,), max_iter=300, random_state=42)
mlp.fit(X_train, y_train)
y_pred = mlp.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy of the one hidden layer MLP model: {accuracy}")

Accuracy of the one hidden layer MLP model: 0.9614285714285714


### To report the performance:

- In order to predict accuracy of the model based on the training dataset, y_pred is calculated using mlp.predict(X_test).
- The accuracy is found using the accuracy_score function.

The accuracy is said to be the proportion of the total number of predictions that were correct. A high accuracy indicates that the model is making predictions that closely match the true labels for most of the inputs. Here it is said to be 96.14% on the testing data.

# 3. From question 2, set the number of hidden layers of the MLP model as [2,4,6,8,10], set the hidden layer size as 100, show the accuracies on the test set.


#### The ideal number of hidden layers depends on the specific task and the nature of the data. Some tasks may require only a few layers, while others may benefit from deeper architectures. Changing the number of hidden layers in a Multi-Layer Perceptron (MLP) model can significantly impact the model's ability to learn and generalize from the data, thereby affecting its accuracy. With more hidden layers, an MLP can model more complex non-linear relationships between the input features and the output, which might not be possible with a shallower network.  When the right number of hidden layers is used, the model is more likely to generalize well to new, unseen data, improving its accuracy on the test set. As the number of layers increases, especially in very deep networks, issues like vanishing or exploding gradients can occur during training, making it harder for the model to converge to an optimal solution.

### MLP Classifier has several parameters.

MLPClassifier(hidden_layer_sizes=(100,), activation='relu', *, solver='adam', alpha=0.0001, batch_size='auto', learning_rate='constant', learning_rate_init=0.001, power_t=0.5, max_iter=200, shuffle=True, random_state=None, tol=0.0001, verbose=False, warm_start=False, momentum=0.9, nesterovs_momentum=True, early_stopping=False, validation_fraction=0.1, beta_1=0.9, beta_2=0.999, epsilon=1e-08, n_iter_no_change=10, max_fun=15000)

#### Here the max_iter is changed to 300 and random_state is set to 42. Other parameters are set to their default values.

### Here the number of layers, n_layers - 2,4,6,8,10 and hidden_layer_sizes is set to 100.

In [7]:
hidden_layers = [2, 4, 6, 8, 10]
accuracies = []
for layers in hidden_layers:
    mlp = MLPClassifier(hidden_layer_sizes=(100,) * layers, max_iter=300, random_state=42)
    mlp.fit(X_train, y_train)
    y_pred = mlp.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    accuracies.append(accuracy)
    print(f"Accuracy with {layers} hidden layers: {accuracy}")

Accuracy with 2 hidden layers: 0.9702142857142857
Accuracy with 4 hidden layers: 0.9694285714285714
Accuracy with 6 hidden layers: 0.9742857142857143
Accuracy with 8 hidden layers: 0.9732857142857143
Accuracy with 10 hidden layers: 0.9757142857142858


The accuracy is said to be the proportion of the total number of predictions that were correct. A high accuracy indicates that the model is making predictions that closely match the true labels for most of the inputs. 

### Accuracy based on number of hidden layers:

1. Accuracy with 2 hidden layers: 97%
2. Accuracy with 4 hidden layers: 96.9%
3. Accuracy with 6 hidden layers: 97.4%
4. Accuracy with 8 hidden layers: 97.32%
5. Accuracy with 10 hidden layers: 97.57%

# 4. From question 2, for the hidden layer set the hidden layer size as [50, 100, 150, 200], show the accuracies on the test set.

#### The optimal size of the hidden layers is often determined through experimentation. It involves trying different sizes and evaluating how they affect model accuracy. Changing the size of hidden layers in a Multi-Layer Perceptron (MLP) model can significantly influence the model's ability to learn from the data and thus affect its accuracy. Increasing the number of neurons (units) in a hidden layer enhances the model's capacity to learn and represent more complex functions. More neurons allow the network to capture a greater variety of patterns and features from the data.  Larger hidden layers can represent more intricate non-linear relationships between the input features and the target output. Just like with adding more hidden layers, increasing the size of hidden layers can lead to overfitting, especially if the dataset is small or noisy. Increasing the size of hidden layers in an MLP model generally enhances the model's ability to learn complex patterns and improves accuracy, especially for challenging tasks. 

### MLP Classifier has several parameters.

MLPClassifier(hidden_layer_sizes=(100,), activation='relu', *, solver='adam', alpha=0.0001, batch_size='auto', learning_rate='constant', learning_rate_init=0.001, power_t=0.5, max_iter=200, shuffle=True, random_state=None, tol=0.0001, verbose=False, warm_start=False, momentum=0.9, nesterovs_momentum=True, early_stopping=False, validation_fraction=0.1, beta_1=0.9, beta_2=0.999, epsilon=1e-08, n_iter_no_change=10, max_fun=15000)

#### Here the max_iter is changed to 300 and random_state is set to 42. Other parameters are set to their default values.

### Here the number of layers, n_layers - 2 and hidden_layer_sizes is set to 50, 100, 150, 200.

In [8]:
hidden_layer_sizes = [50, 100, 150, 200]
accuracies_size = []
for size in hidden_layer_sizes:
    mlp = MLPClassifier(hidden_layer_sizes=(size,), max_iter=300, random_state=42)
    mlp.fit(X_train, y_train)
    y_pred = mlp.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    accuracies_size.append(accuracy)
    print(f"Accuracy with hidden layer size {size}: {accuracy}")

Accuracy with hidden layer size 50: 0.9545
Accuracy with hidden layer size 100: 0.9614285714285714
Accuracy with hidden layer size 150: 0.9685714285714285
Accuracy with hidden layer size 200: 0.9684285714285714


The accuracy is said to be the proportion of the total number of predictions that were correct. A high accuracy indicates that the model is making predictions that closely match the true labels for most of the inputs. 

### Accuracy based on hidden layer sizes:

1. Accuracy with 50 hidden layer sizes: 95.45%
2. Accuracy with 100 hidden layer sizes: 96.14%
3. Accuracy with 150 hidden layer sizes: 96.85%
4. Accuracy with 200 hidden layer sizes: 96.84%

# 5. Based on question 3 and 4, explain the key findings

In [1]:
print("\nKey Findings:")
print("1. To improve the accuracy of the MLP model, we have to increase the number of hidden layers.")
print("2. To improve the accuracy, we have to increase the number of hidden layer sizes but accuracy may become plateau or decrease after a certain size.")


Key Findings:
1. To improve the accuracy of the MLP model, we have to increase the number of hidden layers.
2. To improve the accuracy, we have to increase the number of hidden layer sizes but accuracy may become plateau or decrease after a certain size.


# REFERENCE:

1. "MLPClassifier," Scikit-learn: Machine Learning in Python, [Online]. Available: https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html

2. "MNIST Dataset," Keras Documentation, [Online]. Available: https://keras.io/api/datasets/mnist/
